### PR TITLE
tests for service were created

### PR DESCRIPTION
**Pull request by oleksiistepaniak**

### CODE REVIEW

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+public class ParserServiceImplTest {
+    private static final List<String> VALID_DATA = List.of("type,fruit,quantity", "b,banana,20",
+            "b,apple,100", "s,banana,100", "p,banana,13", "r,apple,10",
+            "p,apple,20", "p,banana,5", "s,banana,50");
+    private static final List<String> EMPTY_DATA = List.of();
+    private static final List<String> INVALID_DATA = List.of("putin huylo, russia sucks",
+            "glory to Ukraine", "russia, is, a, terrorist state");
```
Comment (VitaliiPotseluiko) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:18]: Are you crazy!!!! You didn't come up something more polite? Change it. It's not seriously.

File: src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import static org.junit.Assert.fail;
+
+import core.basesyntax.service.ReaderService;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ReaderServiceImplTest {
+    private static final String VALID_PATH_TO_FILE = "src/main/resources/input.csv";
```
Comment (VitaliiPotseluiko) [inline src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java:13]:
```suggestion
    private static final String VALID_PATH = "src/main/resources/input.csv";

```

File: src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private static final String INVALID_PATH_TO_FILE = "src/main/resources/invalid.csv";
```
Comment (VitaliiPotseluiko) [inline src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java:14]:
```suggestion
    private static final String INVALID_PATH = "src/main/resources/invalid.csv";

```

File: src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    private static final String EMPTY_DATA_IN_THE_PATH_TO_FILE = "src/main/resources/empty.csv";
```
Comment (VitaliiPotseluiko) [inline src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java:15]:
```suggestion
    private static final String PATH_WITH_EMPTY_DATA = "src/main/resources/empty.csv";

```

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class WriterServiceImplTest {
+    private static final String CORRECT_PATH_TO_FILE = "src/main/resources/output.csv";
```
Comment (VitaliiPotseluiko) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:18]:
```suggestion
    private static final String CORRECT_PATH = "src/main/resources/output.csv";

```

**Review by VitaliiPotseluiko — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java (hunk 5 above)
Comment (VitaliiPotseluiko) [reply src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:18]: And you should make tests for FruitTransaction and all classes in operation pakcage.

**Review by VitaliiPotseluiko — APPROVED** [review]

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            "chemical, addiction, good, boy", "approximately, beautiful, bananas, apples, fruits");
+    private ParserService parserService;
+    private List<FruitTransaction> validDataTransactions;
+    private List<FruitTransaction> emptyDataTransactions;
+
+    @Before
+    public void setUp() {
+        parserService = new ParserServiceImpl();
+        validDataTransactions = parserService.parseData(VALID_DATA);
+        emptyDataTransactions = parserService.parseData(EMPTY_DATA);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:27]: you should not call your test method to prepare data for testing
create transactions list by hands

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    private static final List<String> EMPTY_DATA = List.of();
+    private static final List<String> INVALID_DATA = List.of("cross across, math",
+            "chemical, addiction, good, boy", "approximately, beautiful, bananas, apples, fruits");
+    private ParserService parserService;
+    private List<FruitTransaction> validDataTransactions;
+    private List<FruitTransaction> emptyDataTransactions;
+
+    @Before
+    public void setUp() {
+        parserService = new ParserServiceImpl();
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:25]: now we are creating new ParserServiceImpl() before each test
so let's better move it to BeforeClass

same in all tests

File: src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import core.basesyntax.service.ReaderService;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ReaderServiceImplTest {
+    private static final String VALID_PATH = "src/main/resources/input.csv";
+    private static final String INVALID_PATH = "src/main/resources/invalid.csv";
+    private static final String PATH_WITH_EMPTY_DATA = "src/main/resources/empty.csv";
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/ReaderServiceImplTest.java:15]:
```suggestion
    private static final String VALID_PATH = "src/test/resources/input.csv";
    private static final String INVALID_PATH = "src/test/resources/invalid.csv";
    private static final String PATH_WITH_EMPTY_DATA = "src/test/resources/empty.csv";

```

File: src/test/java/core/basesyntax/service/impl/TransactionHandlerImplTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Before
+    public void setUp() {
+        operationHandlerMap.put(FruitTransaction.Operation.BALANCE, new BalanceOperationHandler());
+        operationHandlerMap.put(FruitTransaction.Operation.SUPPLY, new SupplyOperationHandler());
+        operationHandlerMap.put(FruitTransaction.Operation.PURCHASE,
+                new PurchaseOperationHandler());
+        operationHandlerMap.put(FruitTransaction.Operation.RETURN, new ReturnOperationHandler());
+        operationStrategy = new OperationStrategyImpl(operationHandlerMap);
+        transactionHandler = new TransactionHandlerImpl(operationStrategy);
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/TransactionHandlerImplTest.java:45]: move to BeforeClass

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class WriterServiceImplTest {
+    private static final String CORRECT_PATH = "src/main/resources/output.csv";
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:18]:
```suggestion
    private static final String CORRECT_PATH = "src/test/resources/output.csv";

```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
+    private static final FruitTransaction.Operation BALANCE
+            = FruitTransaction.Operation.BALANCE;
+    private static final FruitTransaction.Operation SUPPLY
+            = FruitTransaction.Operation.SUPPLY;
+    private static final FruitTransaction.Operation PURCHASE
+            = FruitTransaction.Operation.PURCHASE;
+    private static final FruitTransaction.Operation RETURN
+            = FruitTransaction.Operation.RETURN;
```
Comment (Andriy-dz) [inline src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java:24]: why you need this? Optional

File: src/test/java/core/basesyntax/service/impl/TransactionHandlerImplTest.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        validFruitTransactions.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE,
+                TITLE_FOR_APPLES, 20));
+        validFruitTransactions.add(new FruitTransaction(FruitTransaction.Operation.PURCHASE,
+                TITLE_FOR_BANANAS, 5));
+        validFruitTransactions.add(new FruitTransaction(FruitTransaction.Operation.SUPPLY,
+                TITLE_FOR_BANANAS, 50));
+    }
+
+    @Test (expected = RuntimeException.class)
+      public void processData_emptyListOfData_notOk() {
```
Comment (Andriy-dz) [inline src/test/java/core/basesyntax/service/impl/TransactionHandlerImplTest.java:66]: don`t fixed

File: src/test/java/core/basesyntax/strategy/OperationStrategyImplTest.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        operations.put(FruitTransaction.Operation.SUPPLY, new SupplyOperationHandler());
+        operations.put(FruitTransaction.Operation.PURCHASE, new PurchaseOperationHandler());
+        operations.put(FruitTransaction.Operation.RETURN, new ReturnOperationHandler());
+        operationStrategy = new OperationStrategyImpl(operations);
+    }
+
+    @Test
+    public void get_balanceOperation_ok() {
+        OperationHandler expected = operationStrategy.get(BALANCE_OPERATION);
+        OperationHandler actual = operationStrategy.get(BALANCE_OPERATION);
```
[the remark this answers is not in the document]
Comment (Andriy-dz) [reply src/test/java/core/basesyntax/strategy/OperationStrategyImplTest.java:42]: don`t fixed

**Review by Andriy-dz — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/impl/TransactionHandlerImplTest.java (hunk 12 above)
Comment (oleksiistepaniak) [reply src/test/java/core/basesyntax/service/impl/TransactionHandlerImplTest.java:66]: Why do I need to fix this test? It works correctly. I check here the work of the method for a passed empty input list. If the input list is empty, you need to throw Runtime Exception.

File: src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java (hunk 11 above)
Comment (oleksiistepaniak) [reply src/test/java/core/basesyntax/service/impl/ParserServiceImplTest.java]: In my new commit, I initialized these constants using Optional, but does it make sense?

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test
+    public void writeReportToFile_defaultCase_ok() {
+        String pathToFile = CORRECT_PATH;
+        String report = FULL_REPORT;
+        writerService.writeReportToFile(pathToFile, report);
+        File createdFile = new File(CORRECT_PATH);
+        try {
+            String expected = CORRECT_RESULT_FOR_FULL_REPORT;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:53]:
```suggestion
            String expected = FULL_REPORT;
```

File: src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test
+    public void writeReportToFile_defaultCaseWithEmptyReport_ok() {
+        String pathToFile = CORRECT_PATH;
+        String report = EMPTY_REPORT;
+        writerService.writeReportToFile(pathToFile, report);
+        File createdFile = new File(CORRECT_PATH);
+        try {
+            String expected = CORRECT_RESULT_FOR_EMPTY_REPORT;
```
Comment (KseniiaMakarova) [inline src/test/java/core/basesyntax/service/impl/WriterServiceImplTest.java:69]: same

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**